libr/debug/p/native/linux/linux_debug.c

```diff
@@ -517,3 +521,1439 @@ RList *linux_desc_list (int pid) {
+	char file[128];
+	int len;
+	pid_t mypid;
+	char *buffer;
+	char *pfname;
+	char *ppsargs;
+	char *basename; /* pr_fname stores just the exec, withouth the path */
+	const char prog_states[] = "RSDTZW";				/* fs/binfmt_elf.c from kernel */
+	prpsinfo_t *p;
+
+	pfname = ppsargs = basename = NULL;
+
+	p = R_NEW0 (prpsinfo_t);
+	if (!p) {
+		eprintf ("Couldn't allocate memory for prpsinfo_t\n");
+		return NULL;
+	}
+
+	p->pr_pid = mypid = dbg->pid;
+
+	/* Start filling pr_fname and pr_psargs */
+	snprintf (file, sizeof(file), "/proc/%d/cmdline", mypid);
+	eprintf ("r_file_slurp\n");
+	buffer = r_file_slurp (file, &len);
+	if (!buffer) {
+		eprintf ("buffer NULL\n");
+		return NULL;
+	}
+	
+	buffer = r_file_slurp (file, &len);
+	pfname = prpsinfo_get_fname (buffer);	
+	if (!pfname) {
+		eprintf ("prpsinfo_get_fname: couldn't allocate memory\n");
+		goto error;
+	}
+	
+	basename = get_basename (pfname, strlen(pfname));
+	strncpy (p->pr_fname, basename, sizeof(p->pr_fname));
+
+	ppsargs = prpsinfo_get_psargs (buffer, pfname, sizeof(p->pr_psargs), len);
+	if (!ppsargs) {
+		eprintf ("prpsinfo_get_psargs: couldn't allocate memory\n");
+		return NULL;
+	}
+
+	strncpy (p->pr_psargs, ppsargs, sizeof(p->pr_psargs));
+	free (buffer);
+	free (ppsargs);
+	free (pfname);
+	
+	p->pr_sname = proc_data->s_name;
+	p->pr_zomb = (p->pr_sname == 'Z') ? 1 : 0;
+	p->pr_state = strchr (prog_states, p->pr_sname) - prog_states;
+	p->pr_ppid = proc_data->ppid;
+	p->pr_pgrp = proc_data->pgrp;
+	p->pr_sid = proc_data->sid;
+	p->pr_flag = proc_data->flag;
+	p->pr_nice = proc_data->nice;
+	p->pr_uid = proc_data->uid;
+	p->pr_gid = proc_data->gid;
+	
+	debug_print_prpsinfo(p);
+	eprintf ("linux_get_prpsinfo: end\n");
+
+	return p;
+
+	error:
+		free (p);
+		free (buffer);
+		free (pfname);
+		free (ppsargs);
+
+		return NULL;
+}
+
+void debug_prstatus(prstatus_t *p) {
+	eprintf("\n== debug_prstatus ==\n");
+	eprintf("p->pr_cursig: %d\n", p->pr_cursig);
+	eprintf("p->pr_info.si_signo: %d\n", p->pr_info.si_signo);
+	eprintf("p->pr_pid: %d\n", p->pr_pid);
+	eprintf("p->pr_ppid: %d\n", p->pr_ppid);
+	eprintf("p->pr_pgrp: %d\n", p->pr_pgrp);
+	eprintf("p->pr_sid: %d\n", p->pr_sid);
+	eprintf("p->pr_sigpend: %ld\n", p->pr_sigpend);
+	eprintf("p->pr_sighold: %ld\n", p->pr_sighold);
+}
+
+static prstatus_t *linux_get_prstatus(RDebug *dbg, proc_stat_content_t *proc_data, short int signr) {
+	prstatus_t *p;
+	ut8 *reg_buff;
+	int rbytes;
+	size_t size_gp_regset;
+
+	reg_buff = R_NEW0 (struct user_regs_struct);
+	if (!reg_buff)
+		return NULL;
+
+	size_gp_regset = sizeof (elf_gregset_t);
+	rbytes = linux_reg_read (dbg, R_REG_TYPE_GPR, reg_buff, size_gp_regset);
+	if (rbytes != size_gp_regset) {                    /* something went wrong */
+		printf ("linux_get_prstatus: error in (rbytes != size_gp_regset)\n");
+		goto error;
+	}
+
+	/* http://lxr.free-electrons.com/source/arch/x86/include/asm/signal.h#L24 */
+	
+	p = R_NEW0 (prstatus_t);
+	if (!p) goto error;
```

Comment:
Avoid gotos unless the function flow is too complex. You can move all the error code in here